types/underscore/index.d.ts

```diff
@@ -176,45 +210,52 @@ declare module _ {
             context?: any): _.Dictionary<T>;
 
         /**
-        * Produces a new array of values by mapping each value in list through a transformation function
-        * (iterator). If the native map method exists, it will be used instead. If list is a JavaScript
-        * object, iterator's arguments will be (value, key, object).
-        * @param list Maps the elements of this array.
-        * @param iterator Map iterator function for each element in `list`.
-        * @param context `this` object in `iterator`, optional.
-        * @return The mapped array result.
-        **/
+         * Produces a new array of values by mapping each value in the collection through a transformation function
+         * (iterator). Each invocation of iterator is called with three arguments: (value, key, collection).
+         * @param collection Maps the elements of this collection.
+         * @param iterator Map iterator function for each element in the collection.
+         * @param context `this` object in `iterator`, optional.
+         * @returns The mapped result.
+         **/
         map<T, TResult>(
-            list: _.List<T>,
-            iterator: _.ListIterator<T, TResult>,
+            collection: List<T>,
+            iterator: ListIterator<T, TResult>,
             context?: any): TResult[];
-
-        map<T>(
-            list: _.List<T>,
-            iterator: _.IterateePropertyShorthand,
-            context?: any): any[];
-
-        map<T>(
-            list: _.List<T>,
-            iterator: _.IterateeMatcherShorthand<any>,
-            context?: any): boolean[];
-
-        /**
-        * @see _.map
-        * @param object Maps the properties of this object.
-        * @param iterator Map iterator function for each property on `object`.
-        * @param context `this` object in `iterator`, optional.
-        * @return The mapped object result.
-        **/
         map<T, TResult>(
-            object: _.Dictionary<T>,
-            iterator: _.ObjectIterator<T, TResult>,
+            collection: Dictionary<T>,
+            iterator: ObjectIterator<T, TResult>,
             context?: any): TResult[];
 
         /**
-        * @see _.map
-        **/
-        collect: typeof _.map;
+         * Produces a new array of values by retrieving the value of the specified property from each item
+         * in the collection.
+         * @param collection The collection of items.
+         * @param iterator The name of a specific property to retrieve from all items.
+         * @returns The set of values for the specified property for each item in the collection.
+         */
+        map<T, K extends KeysOfUnion<T>>(
+            collection: Collection<T>,
+            iterator: K): TypesOfUnionProperty<T, K>[];
```

Comment:
I did some interesting tricky stuff around being able to arrive at the correct set of possible types when iterating over a collection of a type union (e.g. `map([{ a: 'a' }, { a: true }], 'a']) => (string | boolean)[]`), but with the generic `(iterator: string) => any[]` fallback overload that I later put in place to avoid re-introducing #33479 it's probably worth discussing whether this is overkill where the harm to performance will outweigh the help it may provide in what is likely an edge case. I could instead change this overload to `<K extends keyof T>(iterator: K) => T[K]` and let type unions fall back to the `(iterator: string) => any[]` overload, which will probably improve performance to some degree. I'm admittedly not familiar with benchmarking TS definitions, but I could also look into providing a more concrete performance analysis here if folks would like.

Reply:
Not that I really trust this much at this point, but I guess typescript-bot doesn't think that this has a negative impact, so I guess maybe it's alright from a performance perspective. It still might not be great from a readability perspective, though, since it involves some complicated conditional typing.

Reply:
I like the idea, and I'll give it a go, but something I ran into when even trying to use `CollectionIterator` in more places is that type inference for function generics regrettably currently doesn't work well when TS is presented with options on possible interpretations of something like an iterator, so it may not be possible to do this at this time and still have TS automatically infer function type arguments like `T` and `TResult` in `UnderscoreStatic.map<T, TResult>`.

Reply:
This worked better than I thought it would when trying it out, but surprisingly it fell apart on property name iterators; it seems like TS might not be willing to infer a constant string type except where constraints like `extends keyof T` force it to. What I tried is below.

```TS
    type IterateeResult<V, I, T extends TypeOfCollection<V> = TypeOfCollection<V>> = I extends void
        ? T // default iteratee is _.identity
        : I extends keyof T ? T[I]
        : I extends string ? any
        : I extends CollectionIterator<T, infer R, V> ? R
        : I extends Array<string | number> ? any
        : I extends Partial<T> ? boolean
        : never;

        /**
         * Produces a new array of values by mapping each value in the collection through a transformation function
         * (iterator). Each invocation of iterator is called with three arguments: (value, key, collection).
         * @param collection Maps the elements of this collection.
         * @param iterator Map iterator function for each element in the collection.
         * @param context `this` object in `iterator`, optional.
         * @returns The mapped result.
         **/
        map<V extends Collection<any>, I>(
            collection: V,
            iterator: I,
            context?: any): IterateeResult<V, I>[];

    // function iterator - lists
    _.map(stringRecordList, stringRecordListValueIterator); // $ExpectType string[] PASS
    // function iterator - dictionaries
    _.map(stringRecordDictionary, stringRecordDictionaryValueIterator); // $ExpectType string[] PASS
    // function iterator - strings
    _.map(simpleString, stringListValueIterator); // $ExpectType number[] PASS
    // partial object iterator - lists
    _.map(stringRecordList, partialStringRecord); // $ExpectType boolean[] PASS
    // partial object iterator - dictionaries
    _.map(stringRecordDictionary, partialStringRecord); // $ExpectType boolean[] PASS
    // partial object iterator with type any (see #33479)
    _.map(anyValue, partialStringRecord); // $ExpectType boolean[] PASS
    // property name iterator with a non-nullable single type - lists
    _.map(stringRecordList, stringRecordProperty); // $ExpectType string[] FAIL actual any[] - I is inferred as string rather than 'a'
    // property name iterator with a non-nullable single type - dictionaries
    _.map(stringRecordDictionary, stringRecordProperty); // $ExpectType string[] FAIL actual any[] - I is inferred as string rather than 'a'
    // property name iterator with other types - lists
    _.map(stringRecordOrUndefinedList, stringRecordProperty); // $ExpectType any[] PASS
```

Also, I'm pretty sure this approach requires switching to an `UnderscoreStatic` implementation that uses collection-based generics rather than collection-item-based ones to end up with the correct `V` for iterators, and while I was considering switching to that long-term I wasn't originally planning to as part of this PR set. Do you think switching to a `V` generic rather than `T` in `UnderscoreStatic` functions would be a good change to make sooner rather than later? If so, I might consider also switching `Underscore<T, V>`  to `Underscore<V, T extends TypeOfCollection<T> = TypeOfCollection<T>>` (and the same for `_Chain`) in a subsequent PR rather than keeping them in an order that I consider to be backwards until the end of this set of PRs.

Reply:
As a side note, `stringRecordProperty` is defined as `const stringRecordProperty = 'a';` which causes it to have type `'a'`, but just to be extra thorough I tried hard-coding `'a'` as the iteratee parameter for map and ended up with the same failed result 😞.

Reply:
I guess overall I think that I'd prefer to get these changes and tests in place with generics more or less as they currently are (largely `T`-centric rather than `V`-centric) and then try to do more complicated sweeping generics changes like this one all in one go and use the tests to help ensure that everything still works. What do you think?

Reply:
Alright, I was super curious and tried constraining `I` with all of its possible types and found that all of my tests passed if I did that, so this is something that I could move forward with sooner rather than later if you'd like.

```TS
    type IterateeTypes<T> = Function | keyof T | string | Partial<T> | Array<string | number> | undefined;

    type IterateeResult<V, I, T extends TypeOfCollection<V> = TypeOfCollection<V>> = I extends undefined
        ? T // default iteratee is _.identity
        : I extends keyof T ? T[I]
        : I extends string ? any
        : I extends CollectionIterator<T, infer R, V> ? R
        : I extends Array<string | number> ? any
        : I extends Partial<T> ? boolean
        : never;

        /**
         * Produces a new array of values by mapping each value in the collection through a transformation function
         * (iterator). Each invocation of iterator is called with three arguments: (value, key, collection).
         * @param collection Maps the elements of this collection.
         * @param iterator Map iterator function for each element in the collection.
         * @param context `this` object in `iterator`, optional.
         * @returns The mapped result.
         **/
        map<V extends Collection<any>, I extends IterateeTypes<TypeOfCollection<V>>>(
            collection: V,
            iterator: I,
            context?: any): IterateeResult<V, I>[];
```

I might mess around with this a bit more to see if it's possible to use a better function iterator constraint than `Function`.

Reply:
I'm admittedly not sure how much weight this will carry with you, but thinking through this a little more I guess one thing I like about having multiple overloads rather than doing something like this is that some IDEs (VS at least) then shows that there are multiple options and provides a nicer view of what those options are. Even without VS Intellisense (or the equivalent in other IDEs), if I weren't me and I wanted to figure out what types of iterators Underscore functions would take and what the result would be, `IterateeResult` would be a lot harder for me to read than overloads.

![image](https://user-images.githubusercontent.com/8692126/86434852-60f41c00-bcb3-11ea-9794-7b45ccc1be86.png)
![image](https://user-images.githubusercontent.com/8692126/86434865-6c474780-bcb3-11ea-8b62-cb60922290d0.png)
![image](https://user-images.githubusercontent.com/8692126/86434889-7c5f2700-bcb3-11ea-8b9c-323c405a5284.png)
![image](https://user-images.githubusercontent.com/8692126/86434967-add7f280-bcb3-11ea-82d9-fb654d612355.png)
![image](https://user-images.githubusercontent.com/8692126/86434901-854ff880-bcb3-11ea-9fb9-9db238afee75.png)

Also aargh, I thought that summary comments would apply themselves to multiple overloads, but apparently I was wrong about that 😞. I'm glad this comment thread caused me to check that, though.

Reply:
Hmm, also now annoyed at VS for not doing as good a job on overload 1 which has summary comments and is nearly identical to overload 2 except that overload 2 does not have summary comments.

Having experimented a bit more, I've found that VS specifically does worst with the first overload (if I move overloads 3 and 4 above 1 and 2 then it starts throwing a bunch of `unknown`s in overload 3 instead and overload 1 looks nicer with its preserved generic placeholders).

Reply:
The correct function constraint for `IterateeTypes` is `CollectionIterator<T, any, Collection<T>>`. Though I think you never really need `IterateeTypes`, since `IterateeResult` already ensures that only valid iteratees typecheck.

As for you other question, in my opinion the code is much more important than the IDE. IDEs should adapt to work well with our code, not the other way round.

Reply:
By the way, how about `DeepProperty`? Didn't seem feasible?

Reply:
Somehow I missed a few comments of you in the middle, which explains why I think you don't need to constrain the iteratee type even though you explained why you do. I agree with a change like this in a separate PR instead.

Reply:
Unless you see no problem with adding it now, because I do prefer having it sooner than later.

Reply:
I thought I might run into issues with automatic inference of types for `CollectionIterator` parameters (which is admittedly something my tests don't do a good job of testing since they declare the callbacks with types already in place on parameters), but I didn't, so as far as I can tell there aren't any issues with this implementation. While this isn't technically a type union in letter, I think that consolidating this into a single type does follow the spirit of the [Do's and Don'ts Union Types Guidelines](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#use-union-types) of making it so that as many different types can be passed along through call chains as possible, so I think it would be good to add this now. As an added bonus, if everything keeps the `context` parameter I can test for its existence for iterator calls and test for it's optionalness in all other calls, so this will also give me a way to collapse those variations down that I'm happier about.

I skipped over checking the more complex `DeepProperty` type for basic vetting, but I'll check that out now and see if it introduces any issues. It will definitely require a bump up to TS3.0 to be able to use, and it might even need 3.1; I'll let you know what I find out.

Reply:
I can't find a way around TS not liking the possibility of infinite recursion with `DeepProperty` referring to itself, and even if I were to limit the number of possible iterations by declaring that type multiple times and chaining them together I think we'd probably still need the stuff from [Variadic tuple types](https://github.com/microsoft/TypeScript/pull/39094) that will be included in TS4.0 for `[infer Head, infer ...Tail]` to work properly.

Even then, I think TS is likely to infer `string[]` for `I` rather than a tuple type like `['property1', 'property2', ...]`, and I'm not sure that there's a way to constrain it to not do that (`[keyof T, ...Array<string | number | symbol>]` might help a little, but I'm not sure if that would work for anything beyond the first parameter). That wouldn't make this useless, but one would then be required to specify the generic arguments when making a call like this, which is not ideal.

Reply:
I've applied the changes you recommended to use `Iteratee` and `IterateeResult` types in `map`, `filter`, and `groupBy`; let me know what you think. Since I'm switching functions that take iteratees over to collection-based generics (`V` rather than `T`) as part of this, let me know what you think about also using collection-based generics for consistency in functions like `pluck` that don't necessarily require them.

Reply:
If it would be preferable to not switch these generics in `UnderscoreStatic`, I think I could expand this to two separate overloads, one that covers `List` and one that covers `Dictionary`. However, I feel like getting the right `V` for iteratee callbacks is nice and the addition of `I` will already break anyone who is currently explicitly specifying these generics, so now would probably be a good time to make this generics change anyway.

Reply:
I'm fine with a wholesale switch to collection-based generics (so including `pluck`).

As for `DeepProperty`, yeah, that's the part I doubted.

Reply:
Sounds good, made the updates.

Reply:
> While this isn't technically a type union in letter, I think that consolidating this into a single type does follow the spirit of the [Do's and Don'ts Union Types Guidelines](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#use-union-types) of making it so that as many different type possibilities can be passed along through call chains as possible, so I think it would be good to add this now.

Just wanted to note that my statement above was incorrect; `Iteratee` is absolutely a type union literally as well as spiritually. (I think I was thinking about the conditional type `IterateeResult` when I made that comment.)